cpp/src/arrow/array/concatenate.cc

```diff
@@ -243,35 +243,57 @@ class ConcatenateImpl {
  private:
   // Gather the index-th buffer of each input into a vector.
   // Bytes are sliced with that input's offset and length.
+  // Note that BufferVector will not contain the buffer of in_[i] if it's
+  // nullptr.
   BufferVector Buffers(size_t index) {
-    BufferVector buffers(in_.size());
+    BufferVector buffers;
+    buffers.reserve(in_.size());
     for (size_t i = 0; i < in_.size(); ++i) {
-      buffers[i] = SliceBuffer(in_[i].buffers[index], in_[i].offset, in_[i].length);
+      const auto& buffer = in_[i].buffers[index];
+      if (buffer) {
```

Comment:
Does this change work when appending some vectors with a null-bitmap present and some not-present (i.e. 0 nulles) or is index always  > 0?  Could you add this as a to your test case to confirm?

Reply:
Ah, I see it is only called with index >0